### PR TITLE
Create Item::full_name method to include name of reexports

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
I originally wanted to make `Item::name` field private and implement `full_name` as `name` replacement directly. Not sure if it's a good idea though. What do you think?

cc @camelid 
r? @notriddle

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Except that in some cases, we filter on `Item.name.is_none`, so it would be a problem to always return the `name` of the reexport. I can still make the field private and provide both methods though.

**Comment by GuillaumeGomez** [comment]
If an `Item` doesn't have a `name`, it doesn't have a page for itself and it's how it's filtered in `html/render` and why reexports don't have a page on their own.

**Comment by GuillaumeGomez** [comment]
> The item must be a direct descendant of a module; associated items can have names, but still get no pages to themselves.

With this, reexports should get their own page since they are direct descendant of a module (the reexport itself).

**Comment by GuillaumeGomez** [comment]
Hum... I could write a function which handles that I guess. That'd be interesting. :)

**Comment by GuillaumeGomez** [comment]
Absolutely! Let's wait for them then.

**Comment by camelid** [comment]
Sorry, I didn't realize you were waiting on me! I've been busier with other stuff recently. What is it that you wanted my feedback on?

**Comment by GuillaumeGomez** [comment]
We wanted your opinion on "should we add a method which returns `true` if the the item should get its own documentation page instead of basing it on whether it has a name or not". :)

**Comment by GuillaumeGomez** [comment]
Fixed merge conflict.

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -404,6 +404,16 @@ pub(crate) fn rustc_span(def_id: DefId, tcx: TyCtxt<'_>) -> Span {
 }
 
 impl Item {
+    /// If `item.name` is `None` and `self` is an `ImportItem`, it'll look for it.
+    ///
+    /// It's especially usefully to get the name of `pub use x;` or `pub use x as y;`.
+    pub(crate) fn full_name(&self) -> Option<Symbol> {
+        self.name.or_else(|| {
+            if let ImportItem(ref i) = *self.kind &&
+            let ImportKind::Simple(s) = i.kind { Some(s) } else { None }
```
Comment (camelid) [inline src/librustdoc/clean/types.rs:413]: Does `s` refer to the `y` or the `x` in `pub use x as y;`?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:413]: Yes. `Simple` is created with `new_simple` which returns an `Import`. The `Import` struct contains the `source` (so `x` in this case) and the kind (our `Simple("y")`).

**Comment by GuillaumeGomez** [comment]
So currently, the filtering of which item should get its own page is [here](https://github.com/rust-lang/rust/blob/56fd680cf9226ab424f88d4e3b43c5e088d17f19/src/librustdoc/formats/cache.rs#L341-L399). The other place is in [print_sidebar](https://github.com/rust-lang/rust/blob/56fd680cf9226ab424f88d4e3b43c5e088d17f19/src/librustdoc/html/render/mod.rs#L1770-L1780). In both cases, we need to check the item to call the correct function. I can eventually centralize it with a function taking a callback, but I'm not sure we can do much better. At least the logic would be in one place only. What do you think?

**Comment by jyn514** [comment]
It's been a while - @GuillaumeGomez are you planning to follow up on this?

**Comment by GuillaumeGomez** [comment]
No I don't, or at least not before long. Closing then.

**Pull request closed without merging**